plugins/view/src/types.ts

```diff
@@ -0,0 +1,776 @@
+  getAttrClass: () => Ref<Class<Doc>>
+  updateSorting?: (finalOptions: FindOptions<Doc>, attr: AnyAttribute) => Promise<void>
+}
+
+/**
+ * @public
+ */
+export type AggregationManagerResource = Resource<AggregationManager>
+
+/**
+ * @public
+ */
+export type CreateAggregationManagerFunc = Resource<(client: Client, lqCallback: () => void) => AggregationManager>
+
+/**
+ * @public
+ */
+export interface Aggregation extends Class<Doc> {
+  createAggregationManager: CreateAggregationManagerFunc
+}
+
+/**
+ * @public
+ */
+export interface Viewlet extends Doc {
+  attachTo: Ref<Class<Doc>>
+  baseQuery?: DocumentQuery<Doc>
+  descriptor: Ref<ViewletDescriptor>
+  options?: FindOptions<Doc>
+  config: (BuildModelKey | string)[]
+  configOptions?: ViewletConfigOptions
+  viewOptions?: ViewOptionsModel
+  variant?: string
+}
+
+/**
+ * @public
+ */
+export interface ViewletConfigOptions {
+  hiddenKeys?: string[]
+  strict?: boolean
+  extraProps?: Omit<BuildModelKey, 'key'>
+  sortable?: boolean
+}
+
+/**
+ * @public
+ */
+export interface LinkPresenter extends Doc {
+  pattern: string
+  component: AnyComponent
+}
+
+/**
+ * @public
+ *
+ *  "Alt + K" =\> Alt and K should be pressed together
+ *  "J T" - J and then T shold be pressed.
+ */
+export type KeyBinding = string
+
+/**
+ * @public
+ */
+export type ViewActionInput = 'focus' | 'selection' | 'any' | 'none'
+
+/**
+ * @public
+ */
+
+export type ViewActionFunction<T extends Doc = Doc, P = Record<string, any>> = (
+  doc: T | T[] | undefined,
+  evt?: Event,
+  params?: P
+) => Promise<void>
+
+/**
+ * @public
+ */
+export type ViewActionAvailabilityFunction<T extends Doc = Doc> = (doc: T | T[] | undefined) => Promise<boolean>
+
+/**
+ * @public
+ */
+export type ViewAction<P = Record<string, any>> = Resource<ViewActionFunction<Doc, P>>
+
+/**
+ * @public
+ */
+export interface ActionCategory extends Doc, UXObject {
+  // Does category is visible for use in popup.
+  visible: boolean
+}
+
+/**
+ * @public
+ */
+export type ActionGroup = 'create' | 'edit' | 'associate' | 'copy' | 'tools' | 'other' | 'remove'
+
+/**
+ * @public
+ */
+export interface Action<T extends Doc = Doc, P = Record<string, any>> extends Doc, UXObject {
+  // Action implementation details
+  action: Resource<ViewActionFunction<T, P>>
+  // Action implementation parameters
+  actionProps?: P
+
+  // If specified, will show sub menu based on actionPopup/actionProps
+  actionPopup?: AnyComponent
+
+  // If specified, action could be used only with one item selected.
+  // single - one object is required
+  // any - one or multiple objects are required
+  // any - any input is suitable.
+  input: ViewActionInput
+  inline?: boolean
+
+  // Focus and/or all selection document should match target class.
+  target: Ref<Class<Doc>>
+  // Action is applicable only for objects matching criteria
+  query?: DocumentQuery<T>
+
+  // Action is shown only if the check is passed
```

Comment:
We have this one, why we need second mechanism? Let's have just one, more suitable for all cases.

Reply:
OK, removed `ActionWithAvailability` for now